symfony/src/EventListener/RequestListener.php

```diff
@@ -74,9 +88,52 @@ private function checkPageNotices()
 
     private function checkSiteSelect()
     {
-        if (!$this->session->has('site') && !preg_match('/^(\/s)?\/(_profiler|_wdt|cron|admin|help|settings|problem|biobank|review|workqueue)($|\/).*/',
-                $this->request->getPathInfo())) {
-            return new RedirectResponse('/');
+        $hasMultiple = ($this->authorizationChecker->isGranted('ROLE_DASHBOARD') && ($this->authorizationChecker->isGranted('ROLE_USER') || $this->authorizationChecker->isGranted('ROLE_ADMIN') || $this->authorizationChecker->isGranted('ROLE_AWARDEE') || $this->authorizationChecker->isGranted('ROLE_DV_ADMIN')));
+        if ($hasMultiple && $this->session->get('isLoginReturn') && !$this->isUpkeepRoute() && !preg_match('/^(\/s)?\/(splash|_wdt)($|\/).*/', $this->request->getPathInfo())) {
+            return new RedirectResponse('/s/splash');
         }
+
+        if (!$this->session->has('site') && !$this->session->has('awardee') && ($this->authorizationChecker->isGranted('ROLE_USER') || $this->authorizationChecker->isGranted('ROLE_AWARDEE'))) {
+            $user = $this->userService->getUser();
+            if (count($user->getSites()) === 1 && empty($user->getAwardees()) && $this->siteService->isValidSite($user->getSites()[0]->email)) {
+                $this->siteService->switchSite($user->getSites()[0]->email);
+            } elseif (count($user->getAwardees()) === 1 && empty($user->getSites())) {
+                $this->siteService->switchSite($user->getAwardees()[0]->email);
+            } elseif (!preg_match('/^(\/s)?\/(_profiler|_wdt|cron|admin|help|settings|problem|biobank|review|workqueue|site|login|site_select|splash)($|\/).*/',
+                    $this->request->getPathInfo()) && !$this->isUpkeepRoute()) {
+                return new RedirectResponse('/s/site/select');
+            }
+        }
+    }
+
+    private function checkLoginExpired()
+    {
+        // log the user out if their session is expired
+        if ($this->userService->isLoginExpired() && $this->request->attributes->get('_route') !== 'logout') {
+            return new RedirectResponse('/s/logout?timeout=1');
+        }
+    }
+
+    public function onKernelFinishRequest()
+    {
+        if ($this->request && !preg_match('/^(\/s)?\/(login|_wdt)($|\/).*/', $this->request->getPathInfo()) && !$this->isUpkeepRoute() && $this->authorizationChecker->isGranted('IS_AUTHENTICATED_FULLY')) {
```

Comment:
If you make a request to a Symfony route that doesn't exist and you aren't logged in, you end up with an error.

For example, while logged out, go to http://localhost:8080/s/asdf

This throws an error when calling the `isGranted` method:

> The token storage contains no authentication token.